pkg/api/pod/util.go

```diff
@@ -415,6 +415,9 @@ func GetValidationOptionsFromPodSpecAndMeta(podSpec, oldPodSpec *api.PodSpec, po
 				}
 			}
 		}
+
+		// if old spec has used image volume source, we must allow it
+		opts.AllowImageVolumeSource = opts.AllowImageVolumeSource || hasUsedImageVolumeSourceWithPodSpec(oldPodSpec)
```

Comment:
I'm surprised this was being forbidden in validation based on feature enablement (in validation options), rather than being dropped before validation (in strategy PrepareForCreate/PrepareForUpdate) based on feature enablement like most other new fields. 

But regardless of where it is done, it should be considering whether the old object already set the field

Reply:
If the old spec doesn't use the image volume, it will be dropped from the new spec like other features. @liggitt but we forget to consider the old object in validation. 

https://github.com/kubernetes/kubernetes/blob/342da505bdefbd849b808cca3cb76c24a993025f/pkg/api/pod/util.go#L1259-L1280

Reply:
In that case, why do we need AllowImageVolumeSource at all?

Reply:
Sorry, I couldn't get your idea. Without `AllowImageVolumeSource`, when the feature gate is turned off, the existing pod with image volume source won't be updated with an error `must specify a volume type` due to validation for the update operation. @liggitt 

- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L5100
- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4047
- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L4184
- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L407
- https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L735-L748

Reply:
`AllowImageVolumeSource` is used to opt-in to allow `Image` field in `VolumeSource` struct when feature-gate is enabled or old pod is using it.

Reply:
>I'm surprised this was being forbidden in validation based on feature enablement (in validation options), rather than being dropped before validation (in strategy PrepareForCreate/PrepareForUpdate) based on feature enablement like most other new fields.

If an old pod has used it, we will leave it as it is when updating the pod. It follows the same pattern as previous features.

Reply:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L622

Reply:
> Without `AllowImageVolumeSource`, when the feature gate is turned off, the existing pod with image volume source won't be updated with an error `must specify a volume type` due to validation for the update operation. @liggitt

No, an existing pod with image volume will not drop the field - https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L1260

Reply:
Validation should not need to care about the feature gate enablement, only whether the field is present, which it doesn't need AllowImageVolumeSource for.

New pods:
- feature gate on, field persists, validation validates
- feature gate off, field is dropped, validation complains about a missing volume type

Update of a pod that already has an image volume:
- feature gate on or off, field is kept, validation validates

Update of a pod that does not already have an image volume:
- feature gate on, field persists, validation validates
- feature gate off, field is dropped, validation complains about a missing volume type

Reply:
> Validation should not need to care about the feature gate enablement, only whether the field is present, which it doesn't need AllowImageVolumeSource for.

Thanks for your explanation. I got it now. 

What do you think? /cc @deads2k @gnufied @saschagrunert 

It's introduced at https://github.com/kubernetes/kubernetes/pull/125660.

In this PR, if you agree, I will remove the `AllowImageVolumeSource` from `PodValidationOptions`.

Reply:
/cc @deads2k @gnufied @saschagrunert

Reply:
> Was there a reason it wasn't done there like other new feature-gated fields?

Is this code doing the drop ineffective? https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L1284-L1320

It was considered in the original implementation and we decided to keep the drop even though it's not necessarily kinder to pod authors to drop their volume mounts.

Reply:
> Is this code doing the drop ineffective?

No, it's doing the right thing... dropping on create, preserving on update if the field is already present in the old object.

Validation was doing the wrong thing on update, by complaining about the existing field if it was preserved on update. That's what this PR is trying to fix. It could either fix it by replicating the same logic that considers feature enablement AND previous existence in an update, or it could fix it by removing the feature enablement consideration from validation altogether (dropping AllowImageVolumeSource from pod validation options completely)

Reply:
>removing the feature enablement consideration from validation altogether (dropping AllowImageVolumeSource from pod validation options completely) - this is simpler

Oh, sure.  Ok.  The validation gate was likely cruft from the earlier impl where we considered rejecting as kinder to the pod author.

Reply:
> @liggitt Without the AllowImageVolumeSource field, if the validation logic added in the new feature has an unknown bug, it is impossible to avoid the risk by setting the feature gate to false.

We *have* to validate data we allow to persist, regardless of feature gate.

If the validation only kicks in when the image field is present, then turning off the feature gate + drop logic ensures no new creations will have that field.

Reply:
okay, I see. will update